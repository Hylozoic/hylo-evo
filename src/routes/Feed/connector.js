import { connect } from 'react-redux'
import { SAMPLE_FEED_ITEMS } from './sampleData'

function mapStateToProps (state) {
  return {
    feedItems: SAMPLE_FEED_ITEMS
  }
}

function mapDispatchToProps (dispatch, props) {
  return {}
}

export default connect(mapStateToProps, mapDispatchToProps)
