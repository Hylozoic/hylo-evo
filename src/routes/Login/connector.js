import { connect } from 'react-redux'
import { checkLogin, login } from './actions'
export default connect(null, {checkLogin, login})
