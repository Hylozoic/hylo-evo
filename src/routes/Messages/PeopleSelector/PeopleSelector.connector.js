import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'

import getQuerystringParam from 'store/selectors/getQuerystringParam'

import {
  addParticipant,
  removeParticipant,
  fetchContacts,
  fetchRecentContacts,
  holoChatContactsSelector,
  findOrCreateThread,
  holoChatMatchesSelector,
  participantsSelector,
  recentContactsSelector,
  setAutocomplete
} from './PeopleSelector.store'

import fetchPeople from 'store/actions/fetchPeople'
import changeQuerystringParam from 'store/actions/changeQuerystringParam'

export function getParticipantSearch (props, participantsFromStore) {
  const participants = getQuerystringParam('participants', null, props)
  if (participants) {
    return participants
      .split(',')
      .filter(p => !participantsFromStore.find(participant => p === participant))
  }
  return null
}

export function mapStateToProps (state, props) {
  const { holoMode } = props
  // only show recentContacts in Hylo mode
  const recentContacts = holoMode ? [] : recentContactsSelector(state)

  const participants = participantsSelector(state, props)
  return {
    autocomplete: state.autocomplete,
    contacts: holoChatContactsSelector(state, props),
    matches: holoChatMatchesSelector(state, props),
    participants,
    participantSearch: getParticipantSearch(props, participants),
    recentContacts
  }
}

export function mapDispatchToProps (dispatch, props) {
  const { holoMode } = props

  return {
    findOrCreateThread: (participantIds, createdAt) => dispatch(findOrCreateThread(participantIds, createdAt, holoMode)),
    fetchContacts: () => dispatch(fetchContacts(holoMode)),
    fetchPeople: (autocomplete, query, first) => dispatch(fetchPeople(autocomplete, query, first, holoMode)),
    ...bindActionCreators({
      addParticipant,
      removeParticipant,
      changeQuerystringParam,
      fetchContacts,
      fetchRecentContacts,
      setAutocomplete
    }, dispatch)
  }
}

export default connect(mapStateToProps, mapDispatchToProps)
