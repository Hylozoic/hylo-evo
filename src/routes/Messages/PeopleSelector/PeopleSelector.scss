.people-selector {
  display: flex;
  flex: 1;
  flex-direction: column;
  height: 100vh;
}

.autocomplete-control {
  display: flex;
  flex-wrap: wrap;
  flex: 1;
  align-items: flex-end;
  border-bottom: 1px $color-rhino-30 solid;
  padding-bottom: $space-2x;
  margin-right: $space-5x;
  min-height: $space-13x;
}

.autocomplete {
  composes: body-drk-lg from 'css/typography.scss';
  flex: 1;
  border: none;
  outline: none;

  // Yep, believe it or not this is necessary... can't group the selectors!
  &::-webkit-input-placeholder {
    color: $color-rhino-30;
    letter-spacing: 0.4;
  }

  &::-moz-placeholder {
    color: $color-rhino-30;
    letter-spacing: 0.4;
  }

  &:-ms-input-placeholder {
    color: $color-rhino-30;
    letter-spacing: 0.4;
  }

  &:-moz-placeholder {
    color: $color-rhino-30;
    letter-spacing: 0.4;
  }

  &::placeholder {
    color: $color-rhino-30;
    letter-spacing: 0.4;
  }
}

.close-button {
  margin-bottom: $space-4x;
  margin-left: $space-6x;
}

.thread-header {
  composes: header from '../Thread/Thread.scss';
  padding-top: $space-4x;
  padding-bottom: $space-3x;
  height: unset;
}

.message-form {
  composes: message-form from '../MessageForm/MessageForm.scss';
}
