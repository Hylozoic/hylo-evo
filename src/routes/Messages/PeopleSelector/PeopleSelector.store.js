import { createSelector } from 'redux-orm'
import { pick } from 'lodash/fp'

import getMe from 'store/selectors/getMe'
import orm from 'store/models'
import {
  CREATE_MESSAGE,
  FIND_OR_CREATE_THREAD
} from 'store/constants'

export const MODULE_NAME = 'PeopleSelector'

export const FETCH_CONTACTS = 'FETCH_CONTACTS'
export const FETCH_RECENT_CONTACTS = 'FETCH_RECENT_CONTACTS'
export const SET_AUTOCOMPLETE = 'PeopleSelector/SET_AUTOCOMPLETE'
export const ADD_PARTICIPANT = 'PeopleSelector/ADD_PARTICIPANT'
export const REMOVE_PARTICIPANT = 'PeopleSelector/REMOVE_PARTICIPANT'

const findOrCreateThreadQuery =
`mutation ($participantIds: [String]) {
  findOrCreateThread(data: {participantIds: $participantIds}) {
    id
    createdAt
    updatedAt
    participants {
      id
      name
      avatarUrl
    }
  }
}`

export function findOrCreateThread (participantIds, createdAt, holoChatAPI = false, query = findOrCreateThreadQuery) {
  return {
    type: FIND_OR_CREATE_THREAD,
    graphql: {
      query,
      variables: {participantIds, createdAt}
    },
    meta: {
      holoChatAPI,
      extractModel: 'MessageThread'
    }
  }
}

const fetchContactsQuery =
`query PeopleContacts ($first: Int) {
  people (first: $first) {
    items {
      id
      name
      avatarUrl
      memberships (first: 1) {
        id
        community {
          id
          name
        }
      }
    }
  }
}`

export function fetchContacts (holoChatAPI = false, query = fetchContactsQuery, first = 50) {
  return {
    type: FETCH_CONTACTS,
    graphql: {
      query,
      variables: { first }
    },
    meta: {
      holoChatAPI,
      extractModel: 'Person'
    }
  }
}

const fetchRecentContactsQuery =
`query RecentPersonConnections ($first: Int) {
  connections (first: $first) {
    items {
      id
      person {
        id
        name
        avatarUrl
        memberships (first: 1) {
          id
          community {
            id
            name
          }
        }
      }
      type
      updatedAt
    }
  }
}`

export function fetchRecentContacts (query = fetchRecentContactsQuery, first = 20) {
  return {
    type: FETCH_RECENT_CONTACTS,
    graphql: {
      query,
      variables: { first }
    },
    meta: { extractModel: 'PersonConnection' }
  }
}

export function addParticipant (id) {
  return {
    type: ADD_PARTICIPANT,
    payload: id
  }
}

export function removeParticipant (id) {
  return {
    type: REMOVE_PARTICIPANT,
    payload: id
  }
}

export function setAutocomplete (autocomplete) {
  return {
    type: SET_AUTOCOMPLETE,
    payload: autocomplete
  }
}

export function pickPersonListItem (person) {
  return {
    ...pick([ 'id', 'name', 'avatarUrl' ], person.ref),
    community: person.memberships.first()
      ? person.memberships.first().community.name : null
  }
}

export function personListItemSelector (session, participants, currentUser, search = () => true) {
  return session.Person
    .all()
    .filter(p => !participants.includes(p.id))
    .filter(search)
    .filter(p => currentUser ? currentUser.id !== p.id : true)
    .orderBy('name')
    .toModelArray()
    .map(pickPersonListItem)
}

export const holoChatContactsSelector = createSelector(
  orm,
  state => state.orm,
  state => state[MODULE_NAME].participants,
  getMe,
  (_, props) => props.holoMode ? p => p.isHoloData : () => true,
  personListItemSelector
)

export const holoChatMatchesSelector = createSelector(
  orm,
  state => state.orm,
  state => state[MODULE_NAME].participants,
  getMe,
  (state, props) => p => {
    const { autocomplete } = state[MODULE_NAME]
    if (autocomplete) {
      const holoFilter = props.holoMode ? p.isHoloData : true
      return holoFilter && p.name.toLowerCase().includes(autocomplete.toLowerCase())
    }
  },
  personListItemSelector
)

const nameSort = (a, b) => {
  const aName = a.name.toUpperCase()
  const bName = b.name.toUpperCase()
  return aName > bName ? 1 : aName < bName ? -1 : 0
}

export function personConnectionListItemSelector (session, participants) {
  return session.PersonConnection
    .all()
    .toModelArray()
    .map(connection => pickPersonListItem(connection.person))
    .filter(person => !participants.includes(person.id))
    .sort(nameSort)
}

export const recentContactsSelector = createSelector(
  orm,
  state => state.orm,
  state => state[MODULE_NAME].participants,
  personConnectionListItemSelector
)

export function participantsFromStore (state) {
  return state[MODULE_NAME].participants
}

export const participantsSelector = createSelector(
  orm,
  state => state.orm,
  participantsFromStore,
  (session, fromStore) => fromStore.map(id =>
    pick([ 'id', 'name', 'avatarUrl' ], session.Person.withId(id).ref))
)

export const defaultState = {
  participants: []
}

export default function reducer (state = defaultState, action) {
  const { error, payload, type, meta } = action
  if (error) return state

  switch (type) {
    case CREATE_MESSAGE:
      if (meta.forNewThread) {
        return defaultState
      }
      break

    case SET_AUTOCOMPLETE:
      return {
        ...state,
        autocomplete: payload
      }

    case ADD_PARTICIPANT:
      return {
        ...state,
        participants: [ ...state.participants, payload ]
      }

    case REMOVE_PARTICIPANT:
      if (payload) {
        return {
          ...state,
          participants: state.participants.filter(p => p !== payload)
        }
      }
      // No payload? Remove the last element.
      return {
        ...state,
        participants: state.participants.slice(0, state.participants.length - 1)
      }
  }

  return state
}
