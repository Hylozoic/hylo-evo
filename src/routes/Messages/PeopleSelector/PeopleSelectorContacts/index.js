import PeopleSelectorContacts from './PeopleSelectorContacts'

export default PeopleSelectorContacts
