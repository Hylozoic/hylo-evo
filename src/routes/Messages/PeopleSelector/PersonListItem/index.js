import PersonListItem from './PersonListItem'

export default PersonListItem
