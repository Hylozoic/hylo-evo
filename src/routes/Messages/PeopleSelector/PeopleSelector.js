import PropTypes from 'prop-types'
import React from 'react'
import { debounce, throttle } from 'lodash/fp'
import { getKeyCode, keyMap } from 'util/textInput'
import CloseMessages from '../CloseMessages'
import MatchingPeopleList from './MatchingPeopleList'
import PeopleList from './PeopleList'
import MatchingPeopleListItem from './MatchingPeopleListItem'
import './PeopleSelector.scss'

const invalidPersonName = /[^a-z '-]+/gi

export default class PeopleSelector extends React.Component {
  constructor (props) {
    super(props)

    this.state = {
      currentMatch: null
    }
    this.autocompleteInput = React.createRef()
  }

  componentDidMount () {
    this.props.fetchDefaultList()
  }

  componentWillReceiveProps (props) {
    const { matchingPeople } = props
    if (!matchingPeople || matchingPeople.length === 0) {
      this.setCurrentMatch()
      return
    }

    if (matchingPeople.find(m => this.state.currentMatch && m.id === this.state.currentMatch.id)) return
    this.setState({ currentMatch: matchingPeople[0] })
  }

  setCurrentMatch = (person) => this.setState(() => ({ currentMatch: person }))

  selectPerson = (person) => {
    this.autocompleteInput.current.focus()
    if (this.props.selectedPeople.find(p => p.id === person.id)) return
    this.props.setPeopleSearch(null)
    this.setCurrentMatch()
    this.autocompleteInput.current.value = null
    this.props.selectPerson(person)
  }

  removePerson = (person) => {
    this.props.removePerson(person)
  }

  excludeSelectedPeopleAndCurrentUser = (people) => {
    if (!people) return
    const { currentUser, selectedPeople } = this.props
    const selectedPeopleIds = selectedPeople.map(p => p.id)
    return people
      .filter(c => currentUser ? c.id !== currentUser.id : true)
      .filter(c => !selectedPeopleIds.includes(c.id))
  }

  arrow (direction, event) {
    event.preventDefault()
    let delta = 0
    const idx = this.props.matchingPeople.findIndex(m => m.id === this.state.currentMatch.id)
    if (direction === 'up') {
      if (idx > 0) delta = -1
    }
    if (direction === 'down') {
      if (idx < this.props.matchingPeople.length - 1) delta = 1
    }
    this.setCurrentMatch(this.props.matchingPeople[idx + delta])
  }

  autocompleteSearch = throttle(1000, this.props.fetchPeople)

  onChange = debounce(200, () => {
    const { value } = this.autocompleteInput.current
    if (!invalidPersonName.exec(value)) {
      return this.props.setPeopleSearch(value)
    }
    this.autocompleteInput.current.value = value.replace(invalidPersonName, '')
  })

  onKeyDown (evt) {
    switch (getKeyCode(evt)) {
      case keyMap.BACKSPACE: return this.autocompleteInput.current.value ? null : this.removePerson()
      case keyMap.UP: return this.arrow('up', evt)
      case keyMap.DOWN: return this.arrow('down', evt)
      case keyMap.COMMA:
      case keyMap.ENTER:
        evt.preventDefault()
        this.selectPerson(this.state.currentMatch)
        this.autocompleteInput.current.value = null
        return this.props.setPeopleSearch(null)

      default:
        this.autocompleteSearch(this.autocompleteInput.current.value)
    }
  }

  render () {
    const {
      people,
      recentPeople,
      matchingPeople,
      onCloseURL,
      selectedPeople
    } = this.props
    const {
      currentMatch
    } = this.state

    return <React.Fragment>
      <div styleName='thread-header' tabIndex='0'>
        <div styleName='autocomplete-control'>
          {selectedPeople && selectedPeople.map(person =>
            <MatchingPeopleListItem
              avatarUrl={person.avatarUrl}
              name={person.name}
              onClick={() => this.removePerson(person)}
              key={person.id} />
          )}
          <input styleName='autocomplete'
            autoFocus
            ref={this.autocompleteInput}
            type='text'
            spellCheck={false}
            onChange={evt => this.onChange(evt)}
            onKeyDown={evt => this.onKeyDown(evt)}
            placeholder={selectedPeople.length ? '' : 'Type in the names of people to message'} />
        </div>
        <CloseMessages onCloseURL={onCloseURL} />
      </div>
      {currentMatch
        ? <MatchingPeopleList
          matchingPeople={this.excludeSelectedPeopleAndCurrentUser(matchingPeople)}
          currentMatch={currentMatch}
          onClick={this.selectPerson}
          onMouseOver={this.setCurrentMatch} />
        : <PeopleList
          people={this.excludeSelectedPeopleAndCurrentUser(people)}
          recentPeople={this.excludeSelectedPeopleAndCurrentUser(recentPeople)}
          onClick={this.selectPerson} />}
    </React.Fragment>
  }
}

PeopleSelector.propTypes = {
  people: PropTypes.array,
  currentUser: PropTypes.object,
  fetchPeople: PropTypes.func,
  fetchDefaultList: PropTypes.func,
  setPeopleSearch: PropTypes.func,
  matchingPeople: PropTypes.array,
  onCloseURL: PropTypes.string,
  selectedPeople: PropTypes.array,
  recentPeople: PropTypes.array,
  selectPerson: PropTypes.func.isRequired,
  removePerson: PropTypes.func
}
