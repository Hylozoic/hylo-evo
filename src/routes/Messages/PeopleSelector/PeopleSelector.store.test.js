import orm from 'store/models'
import reducer, * as store from './PeopleSelector.store'
import people from './PeopleSelector.test.json'
import { mapStateToProps, getParticipantSearch } from './PeopleSelector.connector'
import {
  CREATE_MESSAGE
} from 'store/constants'

it('matches the last snapshot for findOrCreateThread', () => {
  const graphql = {
    query: 'All the lonely people / Where do they all come from?',
    variables: {
      participantIds: ['1', '2', '3']
    }
  }
  const { query, variables } = graphql
  const actual = store.findOrCreateThread(variables.participantIds, query)
  expect(actual).toMatchSnapshot()
})

it('matches the last snapshot for fetchContacts', () => {
  const graphql = {
    query: 'I see trees of green',
    variables: { first: 20 }
  }
  const { query, variables } = graphql
  const actual = store.fetchContacts(variables.first, query)
  expect(actual).toMatchSnapshot()
})

it('matches the last snapshot for fetchRecentContacts', () => {
  const graphql = {
    query: 'I see them bloom',
    variables: { first: 20 }
  }
  const { query, variables } = graphql
  const actual = store.fetchRecentContacts(variables.first, query)
  expect(actual).toMatchSnapshot()
})

it('returns the correct action from addParticipant', () => {
  const expected = {
    type: store.ADD_PARTICIPANT,
    payload: '1'
  }
  const actual = store.addParticipant('1')
  expect(actual).toEqual(expected)
})

it('returns the correct action from removeParticipant', () => {
  const expected = {
    type: store.REMOVE_PARTICIPANT,
    payload: '1'
  }
  const actual = store.removeParticipant('1')
  expect(actual).toEqual(expected)
})

it('returns the correct action from setAutocomplete', () => {
  const expected = {
    type: store.SET_AUTOCOMPLETE,
    payload: 'flargle'
  }
  const actual = store.setAutocomplete('flargle')
  expect(actual).toEqual(expected)
})

describe('getParticipantSearch', () => {
  it('returns the correct id', () => {
    const location = { search: '?participants=12345' }
    const expected = [ '12345' ]
    const actual = getParticipantSearch({ location }, [])
    expect(actual).toEqual(expected)
  })

  it('returns null if no search', () => {
    const actual = getParticipantSearch({ location: { search: '' } })
    expect(actual).toBe(null)
  })

  it('filters out ids from store', () => {
    const location = { search: '?participants=1,2,3,4' }
    const expected = [ '1', '2', '4' ]
    const actual = getParticipantSearch({ location }, [ '3' ])
    expect(actual).toEqual(expected)
  })
})

describe('connector', () => {
  let session = null
  let state = null

  beforeEach(() => {
    session = orm.mutableSession(orm.getEmptyState())

    people.forEach(person => {
      session.Person.create({
        ...person,
        memberships: person.memberships.map(m => m.id)
      })
      session.Membership.create({
        ...person.memberships[0],
        person: person.id,
        community: person.memberships[0].community.id
      })
      session.Community.create(person.memberships[0].community)
    })
    const me = { id: '999', name: 'Break Wind' }
    session.Person.create(me)
    session.Me.create(me)
    state = { orm: session.state }
    state.PeopleSelector = { participants: people.map(p => p.id) }
  })

  describe('participantsSelector', () => {
    it('returns the correct objects', () => {
      const expected = people.map(p => ({
        id: p.id,
        name: p.name,
        avatarUrl: p.avatarUrl
      }))
      const actual = store.participantsSelector(state)
      expect(actual).toEqual(expected)
    })
  })

  describe('participantsFromStore', () => {
    it('returns the participants array from store', () => {
      const expected = state.PeopleSelector.participants
      const actual = store.participantsFromStore(state)
      expect(actual).toEqual(expected)
    })
  })

  describe('contactsSelector', () => {
    it('filters out currentUser', () => {
      state.PeopleSelector.participants = []
      const actual = store.contactsSelector(state)
      expect(actual.find(p => p.id === '999')).toBe(undefined)
    })
  })

  describe('matchesSelector', () => {
    it('returns empty array if autocomplete is missing', () => {
      const actual = store.matchesSelector(state)
      expect(actual).toEqual([])
    })

    it('returns the correct objects by matching autocomplete', () => {
      state.PeopleSelector.participants = []
      state.PeopleSelector.autocomplete = 'BR'
      const expected = [
        {
          'id': '72203',
          'name': 'Brooks Funk',
          'avatarUrl': 'https://s3.amazonaws.com/uifaces/faces/twitter/matthewkay_/128.jpg',
          'community': 'Associate'
        },
        {
          'id': '72019',
          'name': 'Vita Breitenberg',
          'avatarUrl': 'https://s3.amazonaws.com/uifaces/faces/twitter/markjenkins/128.jpg',
          'community': 'Associate'
        }
      ]
      const actual = store.matchesSelector(state)
      expect(actual).toEqual(expected)
    })

    it('does not return objects whose ids are already in participants', () => {
      state.PeopleSelector.participants = [ '72019' ]
      state.PeopleSelector.autocomplete = 'BR'
      const expected = [
        {
          'id': '72203',
          'name': 'Brooks Funk',
          'avatarUrl': 'https://s3.amazonaws.com/uifaces/faces/twitter/matthewkay_/128.jpg',
          'community': 'Associate'
        }
      ]
      const actual = store.matchesSelector(state)
      expect(actual).toEqual(expected)
    })

    it('filters out currentUser', () => {
      state.PeopleSelector.participants = []
      state.PeopleSelector.autocomplete = 'BRE'
      const expected = [
        {
          'id': '72019',
          'name': 'Vita Breitenberg',
          'avatarUrl': 'https://s3.amazonaws.com/uifaces/faces/twitter/markjenkins/128.jpg',
          'community': 'Associate'
        }
      ]
      const actual = store.matchesSelector(state)
      expect(actual).toEqual(expected)
    })

    describe('pickPersonListItem', () => {
      it('picks the correct properties', () => {
        const person = session.Person.withId('72203')
        const expected = {
          'id': '72203',
          'name': 'Brooks Funk',
          'avatarUrl': 'https://s3.amazonaws.com/uifaces/faces/twitter/matthewkay_/128.jpg',
          'community': 'Associate'
        }
        const actual = store.pickPersonListItem(person)
        expect(actual).toEqual(expected)
      })
    })

    describe('personConnectionListItemSelector', () => {
      let id = 1

      it('filters connections by participants', () => {
        [ '72203', '72019' ].forEach(person =>
          session.PersonConnection.create({
            id: '' + id++,
            person,
            type: 'message'
          }))
        const participants = [ '72203' ]
        const expected = [ '72019' ]
        const actual = store.personConnectionListItemSelector(session, participants)
          .map(person => person.id)
        expect(actual).toEqual(expected)
      })
    })
  })

  describe('mapStateToProps', () => {
    it('returns the correct object', () => {
      state.PeopleSelector.participants = [ '72203' ]
      const actual = mapStateToProps(state, { location: { search: '' } })
      expect(actual).toMatchSnapshot()
    })
  })
})

describe('reducer', () => {
  it('should return the initial state', () => {
    const expected = store.defaultState
    const actual = reducer(undefined, { type: 'FLARGLE' })
    expect(actual).toEqual(expected)
  })

  it('should handle PeopleSelector/ADD_PARTICIPANT', () => {
    const expected = { participants: [ '1' ] }
    const actual = reducer(store.defaultState, {
      type: store.ADD_PARTICIPANT,
      payload: '1'
    })
    expect(actual).toEqual(expected)
  })

  it('should handle PeopleSelector/REMOVE_PARTICIPANT', () => {
    const state = { participants: [ '1', '3', '44444', '123454' ] }
    const expected = { participants: [ '1', '3', '123454' ] }
    const actual = reducer(state, {
      type: store.REMOVE_PARTICIPANT,
      payload: '44444'
    })
    expect(actual).toEqual(expected)
  })

  it('should remove the last participant if REMOVE_PARTICIPANT payload is undefined', () => {
    const state = { participants: [ '1', '3', '44444', '123454' ] }
    const expected = { participants: [ '1', '3', '44444' ] }
    const actual = reducer(state, { type: store.REMOVE_PARTICIPANT })
    expect(actual).toEqual(expected)
  })

  it('should handle PeopleSelector/SET_AUTOCOMPLETE', () => {
    const state = {}
    const expected = { autocomplete: 'flargleargle' }
    const actual = reducer(state, { type: store.SET_AUTOCOMPLETE, payload: 'flargleargle' })
    expect(actual).toEqual(expected)
  })

  it('should return to defaultState if CREATE_MESSAGE happens with forNewThread', () => {
    const expected = store.defaultState
    const initialState = {
      particicipants: [ '1' ],
      autocomplete: 'test'
    }
    const action = {
      type: CREATE_MESSAGE,
      meta: {
        forNewThread: true
      }
    }
    const actual = reducer(initialState, action)
    expect(actual).toEqual(expected)
  })
})
