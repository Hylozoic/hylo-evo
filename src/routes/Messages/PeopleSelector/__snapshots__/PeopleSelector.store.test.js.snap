// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`connector mapStateToProps returns the correct object 1`] = `
Object {
  "autocomplete": undefined,
  "contacts": Array [
    Object {
      "avatarUrl": "https://s3.amazonaws.com/uifaces/faces/twitter/ecommerceil/128.jpg",
      "community": "Associate",
      "id": "71862",
      "name": "Ahmad Bernier",
    },
    Object {
      "avatarUrl": "https://s3.amazonaws.com/uifaces/faces/twitter/akmalfikri/128.jpg",
      "community": "Associate",
      "id": "72655",
      "name": "Leora Bahringer",
    },
    Object {
      "avatarUrl": "https://s3.amazonaws.com/uifaces/faces/twitter/markjenkins/128.jpg",
      "community": "Associate",
      "id": "72019",
      "name": "Vita Breitenberg",
    },
  ],
  "matches": Array [],
  "participantSearch": null,
  "participants": Array [
    Object {
      "avatarUrl": "https://s3.amazonaws.com/uifaces/faces/twitter/matthewkay_/128.jpg",
      "id": "72203",
      "name": "Brooks Funk",
    },
  ],
  "recentContacts": Array [],
}
`;

exports[`matches the last snapshot for fetchContacts 1`] = `
Object {
  "graphql": Object {
    "query": 20,
    "variables": Object {
      "first": "I see trees of green",
    },
  },
  "meta": Object {
    "extractModel": "Person",
    "holochainAPI": true,
  },
  "type": "FETCH_CONTACTS",
}
`;

exports[`matches the last snapshot for fetchRecentContacts 1`] = `
Object {
  "graphql": Object {
    "query": 20,
    "variables": Object {
      "first": "I see them bloom",
    },
  },
  "meta": Object {
    "extractModel": "PersonConnection",
  },
  "type": "FETCH_RECENT_CONTACTS",
}
`;

exports[`matches the last snapshot for findOrCreateThread 1`] = `
Object {
  "graphql": Object {
    "query": "mutation ($participantIds: [String]) {
  findOrCreateThread(data: {participantIds: $participantIds}) {
    id
    createdAt
    updatedAt
    participants {
      id
      name
      avatarUrl
    }
  }
}",
    "variables": Object {
      "createdAt": "All the lonely people / Where do they all come from?",
      "participantIds": Array [
        "1",
        "2",
        "3",
      ],
    },
  },
  "meta": Object {
    "extractModel": "MessageThread",
    "holochainAPI": false,
  },
  "type": "FIND_OR_CREATE_THREAD",
}
`;
